File: tests/ad-hoc-uia.ts
// Ad-hoc test for the UI Automation provider (issue #321, "not accessible to
// screen readers").
//
// Narrator / NVDA read a document through UIA: they look for the Document
// element in the *control* / *content* view, get its TextPattern, and then walk
// a text range with ExpandToEnclosingUnit() + Move() to read character by
// character, word by word, line by line. This test does exactly that from a
// real UIA client (tests/ad-hoc-uia.cpp, compiled on demand with cl.exe) and
// checks that:
//   - the Document element is in the control and content views
//   - its TextPattern returns the document text
//   - walking by character / word / line advances instead of repeating
//   - reading the whole document line by line terminates
//
// It is ad-hoc (not in tests/run-almost-all.ts) because it needs the Visual Studio
// command-line tools to build the client; without cl.exe it skips.

import { existsSync, mkdirSync, rmSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { EXE, ROOT, tmpPath } from "./util";
import {
  moveWindow,
  showWindow,
  sleep,
  waitForChildWindow,
  waitForTopWindow,
  SW_RESTORE,
  killAndWait,
  killProcessesNamed,
} from "./winapi";

const PAGE_LINES = [
  ["Alpha line one", "Bravo line two", "Charlie line three", "Delta line four"],
  ["Echo line five", "Foxtrot line six", "Golf line seven", "Hotel line eight"],
];

// a PDF with the given lines of real text, one page per group
function makeTextPdf(pages: string[][]): Buffer {
  const enc = (s: string) => Buffer.from(s, "latin1");
  const objs: Buffer[] = [];
  const add = (s: string) => {
    objs.push(enc(s));
    return objs.length; // 1-based object number
  };
  add("<< /Type /Catalog /Pages 2 0 R >>"); // 1
  add("PAGES_PLACEHOLDER"); // 2, patched below
  const fontNo = add("<< /Type /Font /Subtype /Type1 /BaseFont /Helvetica /Encoding /WinAnsiEncoding >>");

  const pageNos: number[] = [];
  for (const lines of pages) {
    let stream = "BT /F1 18 Tf 72 700 Td 24 TL\n";
    for (const line of lines) {
      stream += `(${line}) Tj T*\n`;
    }
    stream += "ET";
    const contentNo = add(`<< /Length ${stream.length} >>\nstream\n${stream}\nendstream`);
    pageNos.push(
      add(
        `<< /Type /Page /Parent 2 0 R /MediaBox [0 0 612 792] ` +
          `/Resources << /Font << /F1 ${fontNo} 0 R >> >> /Contents ${contentNo} 0 R >>`,
      ),
    );
  }
  const kids = pageNos.map((n) => `${n} 0 R`).join(" ");
  objs[1] = enc(`<< /Type /Pages /Kids [${kids}] /Count ${pageNos.length} >>`);

  const parts: Buffer[] = [enc("%PDF-1.7\n")];
  const offsets: number[] = [];
  let pos = parts[0]!.length;
  objs.forEach((body, i) => {
    offsets.push(pos);
    const obj = Buffer.concat([enc(`${i + 1} 0 obj\n`), body, enc("\nendobj\n")]);
    parts.push(obj);
    pos += obj.length;
  });
  let xref = `xref\n0 ${objs.length + 1}\n0000000000 65535 f \n`;
  for (const off of offsets) {
    xref += `${String(off).padStart(10, "0")} 00000 n \n`;
  }
  parts.push(enc(`${xref}trailer\n<< /Size ${objs.length + 1} /Root 1 0 R >>\nstartxref\n${pos}\n%%EOF\n`));
  return Buffer.concat(parts);
}

// compile the UIA client; null when the VS command-line tools aren't available
function buildClient(): string | null {
  const src = join(ROOT, "tests", "ad-hoc-uia.cpp");
  const exe = tmpPath("ad-hoc-uia-client.exe");
  const which = Bun.spawnSync(["where", "cl.exe"]);
  if (which.exitCode !== 0) {
    return null;
  }
  const r = Bun.spawnSync(
    [
      "cl.exe",
      "/nologo",
      "/EHsc",
      "/std:c++17",
      src,
      `/Fe${exe}`,
      `/Fo${tmpPath("ad-hoc-uia-client.obj")}`,
      "/link",
      "ole32.lib",
      "oleaut32.lib",
      "user32.lib",
      "uiautomationcore.lib",
    ],
    { cwd: tmpPath("") },
  );
  if (!existsSync(exe)) {
    throw new Error(`could not build the UIA client:\n${r.stdout.toString()}\n${r.stderr.toString()}`);
  }
  return exe;
}

function parseKeyValues(out: string): Map<string, string> {
  const res = new Map<string, string>();
  for (const line of out.split(/\r?\n/)) {
    const i = line.indexOf("=");
    if (i > 0) {
      res.set(
        line.slice(0, i),
        line
          .slice(i + 1)
          .replace(/\\n/g, "\n")
          .replace(/\\p/g, "|"),
      );
    }
  }
  return res;
}

const SETTINGS = [`RestoreSession = false`, `ShowStartPage = false`, `CheckForUpdates = false`, ``].join("\n");

export async function testit(): Promise<void> {
  const client = buildClient();
  if (!client) {
    console.log("  cl.exe not in PATH, skipping the UI Automation test");
    return;
  }

  const pdf = tmpPath("ad-hoc-uia.pdf");
  writeFileSync(pdf, makeTextPdf(PAGE_LINES));
  const appdata = tmpPath("ad-hoc-uia-appdata");
  rmSync(appdata, { recursive: true, force: true });
  mkdirSync(appdata, { recursive: true });
  writeFileSync(join(appdata, "SumatraPDF-settings.txt"), SETTINGS);

  await killProcessesNamed("SumatraPDF.exe");
  const proc = Bun.spawn([EXE, "-for-testing", "-appdata", appdata, pdf], { stdout: "ignore", stderr: "ignore" });
  try {
    const frame = await waitForTopWindow(proc.pid, "SUMATRA_PDF_FRAME");
    if (!frame) {
      throw new Error("SumatraPDF main window did not appear");
    }
    showWindow(frame, SW_RESTORE);
    moveWindow(frame, 60, 60, 1100, 850);
    const canvas = await waitForChildWindow(frame, "SUMATRA_PDF_CANVAS");
    if (!canvas) {
      throw new Error("could not find the canvas window");
    }
    await sleep(1500);

    const r = Bun.spawnSync([client, String(canvas)]);
    const out = r.stdout.toString() + r.stderr.toString();
    const kv = parseKeyValues(out);
    const need = (k: string) => {
      const v = kv.get(k);
      if (v === undefined) {
        throw new Error(`UIA client did not report ${k}:\n${out}`);
      }
      return v;
    };

    if (need("doc.found") !== "1") {
      throw new Error(`no Document element under the canvas:\n${out}`);
    }
    if (need("doc.hasTextPattern") !== "1") {
      throw new Error(`the Document element has no TextPattern:\n${out}`);
    }
    // a Document outside the control / content views is invisible to screen
    // readers even though it holds all the text
    if (need("doc.isControlElement") !== "1" || need("doc.isContentElement") !== "1") {
      throw new Error(
        `Document element is not in the control/content view ` +
          `(isControlElement=${kv.get("doc.isControlElement")}, isContentElement=${kv.get("doc.isContentElement")})`,
      );
    }
    if (need("canvas.kbFocusable") !== "1") {
      throw new Error("the canvas does not report itself as keyboard focusable");
    }

    const firstLine = PAGE_LINES[0]![0]!;
    if (!need("doc.text").startsWith(firstLine)) {
      throw new Error(`document text does not start with '${firstLine}': '${kv.get("doc.text")}'`);
    }

    // each Move() must land on new text, or a screen reader repeats one unit
    // forever (that was the bug: Move() reported success without moving)
    for (const [key, steps] of [
      ["walk.char", 4],
      ["walk.word", 3],
      ["walk.line", 3],
    ] as const) {
      const seen: string[] = [];
      for (let i = 0; i <= steps; i++) {
        seen.push(need(`${key}${i}`));
      }
      const distinct = new Set(seen).size;
      if (distinct <= 1) {
        throw new Error(`walking by ${key} never advanced: ${JSON.stringify(seen)}`);
      }
      if (distinct < seen.length) {
        throw new Error(`walking by ${key} repeated itself: ${JSON.stringify(seen)}`);
      }
    }
    // the second page must be reachable by page
    if (!need("walk.page1").startsWith(PAGE_LINES[1]![0]!)) {
      throw new Error(`moving one page did not land on page 2: '${kv.get("walk.page1")}'`);
    }

    // Reading what is under the mouse / finger (Narrator's mouse mode, touch
    // exploration) needs two things: GetBoundingRectangles to say where the
    // text is on screen, and RangeFromPoint to say what text is at a point.
    // Both used to be E_NOTIMPL.
    if (need("point.haveRect") !== "1") {
      throw new Error(
        `no on-screen rectangle for the first word: GetBoundingRectangles gives screen readers ` +
          `nowhere to point (Narrator can't highlight what it reads)`,
      );
    }
    if (need("point.hr") !== "0x00000000") {
      throw new Error(`RangeFromPoint failed with ${kv.get("point.hr")}: reading text under the mouse can't work`);
    }
    const want = need("point.word").trim();
    const got = need("point.wordAtPoint").trim();
    if (!got.includes(want) && !want.includes(got)) {
      throw new Error(`RangeFromPoint over the word '${want}' returned '${got}'`);
    }

    if (need("sayall.terminated") !== "1") {
      throw new Error("reading the document line by line never reached the end");
    }
    const nLines = parseInt(need("sayall.lines"), 10);
    const totalLines = PAGE_LINES.reduce((n, p) => n + p.length, 0);
    if (nLines < totalLines - 1) {
      throw new Error(`read only ${nLines} lines, expected about ${totalLines}`);
    }
    console.log(`  UIA: document readable, ${nLines} lines walked, page/word/char navigation advances ✓`);
    console.log(`  UIA: word under the mouse: '${kv.get("point.word")}' -> '${kv.get("point.wordAtPoint")}' ✓`);
  } finally {
    await killAndWait(proc);
  }
}

if (import.meta.main) {
  const { runStandalone } = await import("./util");
  await runStandalone(testit);
}

File: tests/ad-hoc-uia.cpp
// UI Automation *client* used by tests/ad-hoc-uia.ts to check what a screen
// reader (Narrator, NVDA, ...) can see and read inside SumatraPDF's canvas.
// It is compiled on demand by that test, not by the normal build:
//   cl /nologo /EHsc /std:c++17 ad-hoc-uia.cpp /Fe<out.exe>
//      /link ole32.lib oleaut32.lib user32.lib uiautomationcore.lib
//
// Usage: ad-hoc-uia.exe <canvas hwnd as decimal>
// Prints `key=value` lines; values escape \ | and newlines.
#include <windows.h>
#include <uiautomation.h>
#include <stdio.h>

static void PrintEscaped(const char* key, BSTR s) {
    printf("%s=", key);
    for (int i = 0; s && s[i]; i++) {
        wchar_t c = s[i];
        if (c == L'\r') {
            continue;
        }
        if (c == L'\n') {
            printf("\\n");
        } else if (c == L'\\') {
            printf("\\\\");
        } else if (c == L'|') {
            printf("\\p");
        } else {
            printf("%lc", c);
        }
    }
    printf("\n");
}

// text of the range, or an empty string
static BSTR RangeText(IUIAutomationTextRange* r, int maxLen) {
    BSTR s = nullptr;
    if (r) {
        r->GetText(maxLen, &s);
    }
    return s;
}

int main(int argc, char** argv) {
    if (argc < 2) {
        printf("error=usage: ad-hoc-uia.exe <canvas hwnd>\n");
        return 1;
    }
    HWND canvas = (HWND)(intptr_t)_atoi64(argv[1]);
    SetConsoleOutputCP(CP_UTF8);
    CoInitializeEx(nullptr, COINIT_MULTITHREADED);

    IUIAutomation* uia = nullptr;
    HRESULT hr =
        CoCreateInstance(__uuidof(CUIAutomation), nullptr, CLSCTX_INPROC_SERVER, __uuidof(IUIAutomation), (void**)&uia);
    if (FAILED(hr) || !uia) {
        printf("error=CoCreateInstance(CUIAutomation) failed 0x%08x\n", (unsigned)hr);
        return 1;
    }

    IUIAutomationElement* canvasElem = nullptr;
    hr = uia->ElementFromHandle(canvas, &canvasElem);
    if (FAILED(hr) || !canvasElem) {
        printf("error=ElementFromHandle failed 0x%08x\n", (unsigned)hr);
        return 1;
    }
    BOOL kbFocusable = FALSE;
    canvasElem->get_CurrentIsKeyboardFocusable(&kbFocusable);
    printf("canvas.found=1\n");
    printf("canvas.kbFocusable=%d\n", (int)kbFocusable);

    // the document element is the canvas child whose control type is Document
    IUIAutomationTreeWalker* raw = nullptr;
    uia->get_RawViewWalker(&raw);
    IUIAutomationElement* doc = nullptr;
    if (raw) {
        IUIAutomationElement* it = nullptr;
        raw->GetFirstChildElement(canvasElem, &it);
        while (it) {
            CONTROLTYPEID ct = 0;
            it->get_CurrentControlType(&ct);
            if (ct == UIA_DocumentControlTypeId) {
                doc = it;
                break;
            }
            IUIAutomationElement* next = nullptr;
            raw->GetNextSiblingElement(it, &next);
            it->Release();
            it = next;
        }
    }
    if (!doc) {
        printf("doc.found=0\n");
        return 0;
    }
    printf("doc.found=1\n");

    BOOL isCtrl = FALSE, isContent = FALSE;
    doc->get_CurrentIsControlElement(&isCtrl);
    doc->get_CurrentIsContentElement(&isContent);
    printf("doc.isControlElement=%d\n", (int)isCtrl);
    printf("doc.isContentElement=%d\n", (int)isContent);
    BSTR name = nullptr;
    doc->get_CurrentName(&name);
    PrintEscaped("doc.name", name);
    if (name) {
        SysFreeString(name);
    }

    IUIAutomationTextPattern* tp = nullptr;
    doc->GetCurrentPatternAs(UIA_TextPatternId, __uuidof(IUIAutomationTextPattern), (void**)&tp);
    printf("doc.hasTextPattern=%d\n", tp ? 1 : 0);
    if (!tp) {
        return 0;
    }

    IUIAutomationTextRange* docRange = nullptr;
    tp->get_DocumentRange(&docRange);
    BSTR all = RangeText(docRange, 400);
    PrintEscaped("doc.text", all);
    if (all) {
        SysFreeString(all);
    }

    // walking by character / word / line is how a screen reader reads: each
    // step must land on new text
    struct {
        TextUnit unit;
        const char* key;
        int steps;
        int maxLen;
    } walks[] = {{TextUnit_Character, "walk.char", 4, 4},
                 {TextUnit_Word, "walk.word", 3, 40},
                 {TextUnit_Line, "walk.line", 3, 80},
                 {TextUnit_Page, "walk.page", 2, 40}};
    for (auto& w : walks) {
        IUIAutomationTextRange* r = nullptr;
        tp->get_DocumentRange(&r);
        if (!r) {
            continue;
        }
        r->MoveEndpointByRange(TextPatternRangeEndpoint_End, r, TextPatternRangeEndpoint_Start);
        r->ExpandToEnclosingUnit(w.unit);
        BSTR s = RangeText(r, w.maxLen);
        char key0[64];
        sprintf_s(key0, "%s0", w.key);
        PrintEscaped(key0, s); // the unit we start on
        if (s) {
            SysFreeString(s);
        }
        for (int i = 1; i <= w.steps; i++) {
            int moved = 0;
            r->Move(w.unit, 1, &moved);
            BSTR s2 = RangeText(r, w.maxLen);
            char key[64];
            sprintf_s(key, "%s%d", w.key, i);
            PrintEscaped(key, s2);
            if (s2) {
                SysFreeString(s2);
            }
        }
        r->Release();
    }

    // say-all: read the whole document line by line, like a screen reader does
    IUIAutomationTextRange* sayAll = nullptr;
    tp->get_DocumentRange(&sayAll);
    if (sayAll) {
        sayAll->MoveEndpointByRange(TextPatternRangeEndpoint_End, sayAll, TextPatternRangeEndpoint_Start);
        sayAll->ExpandToEnclosingUnit(TextUnit_Line);
        const int kMaxLines = 2000;
        int lines = 0;
        for (; lines < kMaxLines; lines++) {
            int moved = 0;
            if (FAILED(sayAll->Move(TextUnit_Line, 1, &moved)) || moved == 0) {
                break;
            }
        }
        printf("sayall.lines=%d\n", lines);
        printf("sayall.terminated=%d\n", lines < kMaxLines ? 1 : 0);
        sayAll->Release();
    }

    // reading what is under the mouse / finger: take the on-screen rectangle of
    // a word, hand its center to RangeFromPoint and see whether we get that
    // same word back
    {
        IUIAutomationTextRange* word = nullptr;
        tp->get_DocumentRange(&word);
        if (word) {
            word->MoveEndpointByRange(TextPatternRangeEndpoint_End, word, TextPatternRangeEndpoint_Start);
            word->ExpandToEnclosingUnit(TextUnit_Word);
            BSTR wantS = RangeText(word, 40);
            PrintEscaped("point.word", wantS);
            SAFEARRAY* rects = nullptr;
            HRESULT hrR = word->GetBoundingRectangles(&rects);
            double cx = 0, cy = 0;
            bool haveRect = false;
            if (SUCCEEDED(hrR) && rects) {
                LONG lo = 0, hi = -1;
                SafeArrayGetLBound(rects, 1, &lo);
                SafeArrayGetUBound(rects, 1, &hi);
                if (hi - lo + 1 >= 4) {
                    double v[4]{};
                    for (LONG i = 0; i < 4; i++) {
                        LONG idx = lo + i;
                        SafeArrayGetElement(rects, &idx, &v[i]);
                    }
                    cx = v[0] + (v[2] / 2);
                    cy = v[1] + (v[3] / 2);
                    haveRect = v[2] > 0 && v[3] > 0;
                }
                SafeArrayDestroy(rects);
            }
            printf("point.haveRect=%d\n", haveRect ? 1 : 0);
            printf("point.x=%d\npoint.y=%d\n", (int)cx, (int)cy);
            if (haveRect) {
                POINT pt{(LONG)cx, (LONG)cy};
                IUIAutomationTextRange* atPt = nullptr;
                HRESULT hrP = tp->RangeFromPoint(pt, &atPt);
                printf("point.hr=0x%08x\n", (unsigned)hrP);
                if (SUCCEEDED(hrP) && atPt) {
                    atPt->ExpandToEnclosingUnit(TextUnit_Word);
                    BSTR gotS = RangeText(atPt, 40);
                    PrintEscaped("point.wordAtPoint", gotS);
                    if (gotS) {
                        SysFreeString(gotS);
                    }
                    atPt->Release();
                }
            }
            if (wantS) {
                SysFreeString(wantS);
            }
            word->Release();
        }
    }

    if (docRange) {
        docRange->Release();
    }
    tp->Release();
    doc->Release();
    if (raw) {
        raw->Release();
    }
    canvasElem->Release();
    uia->Release();
    CoUninitialize();
    return 0;
}
